Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

```diff
@@ -3881,4 +3881,95 @@ HTTPServer server({
+    auto [webView, navigationDelegate] = siteIsolatedViewAndDelegate(configuration);
+
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://example.com/mainframe"]]];
+    [navigationDelegate waitForDidFinishNavigation];
+
+    callMethodOnFirstVideoElementInFrame(webView.get(), @"play", nil);
+    EXPECT_TRUE([webView _isPlayingAudio]);
+
+    callMethodOnFirstVideoElementInFrame(webView.get(), @"play", [webView firstChildFrame]);
+    EXPECT_TRUE([webView _isPlayingAudio]);
+
+    callMethodOnFirstVideoElementInFrame(webView.get(), @"pause", nil);
+    EXPECT_TRUE([webView _isPlayingAudio]) << "Should still be playing audio after pausing in one of the two frames";
+
+    callMethodOnFirstVideoElementInFrame(webView.get(), @"pause", [webView firstChildFrame]);
+    EXPECT_FALSE([webView _isPlayingAudio]) << "Should not be playing audio after pausing in both frames";
+}
+
+TEST(SiteIsolation, PlayAudioInRemoteFrameThenRemove)
+{
+    auto mainFrameHTML = "<iframe id='iframe' src='https://webkit.org/subframe'></iframe>"_s;
+    auto subFrameHTML = "<video src='/video-with-audio.mp4' webkit-playsinline loop></video>"_s;
+
+    RetainPtr<NSData> videoData = [NSData dataWithContentsOfFile:[NSBundle.test_resourcesBundle pathForResource:@"video-with-audio" ofType:@"mp4"] options:0 error:NULL];
+    HTTPResponse videoResponse { videoData.get() };
+    videoResponse.headerFields.set("Content-Type"_s, "video/mp4"_s);
+
+    HTTPServer server({
+        { "/mainframe"_s, { { { "Content-Type"_s, "text/html"_s } }, mainFrameHTML } },
+        { "/subframe"_s, { { { "Content-Type"_s, "text/html"_s } }, subFrameHTML } },
+        { "/video-with-audio.mp4"_s, { videoData.get() } },
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    RetainPtr configuration = server.httpsProxyConfiguration();
+    auto [webView, navigationDelegate] = siteIsolatedViewAndDelegate(configuration);
+
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://example.com/mainframe"]]];
+    [navigationDelegate waitForDidFinishNavigation];
+
+    callMethodOnFirstVideoElementInFrame(webView.get(), @"play", [webView firstChildFrame]);
+    EXPECT_TRUE([webView _isPlayingAudio]);
+
+    __block bool done = false;
+    __block RetainPtr<NSError> error;
+    [webView evaluateJavaScript:@"document.querySelectorAll('iframe').forEach(iframe => iframe.remove())" completionHandler:^(id result, NSError *scriptError) {
+        error = scriptError;
+        done = true;
+    }];
+    TestWebKitAPI::Util::run(&done);
+    EXPECT_FALSE(!!error) << [error description].UTF8String;
+    done = false;
+
+    // Explicitly capture webView since lambdas can't capture a structured binding until c++20.
+    bool isNotPlayingAudio = TestWebKitAPI::Util::waitFor([view = webView]() {
+        return ![view _isPlayingAudio];
+    });
+    EXPECT_TRUE(isNotPlayingAudio) << "Should not be playing audio after removing iframe";
```

Comment:
I suspect these tests might be flaky right now. Instead of waiting for the JavaScript evaluation to finish, you could just wait for the correct state. This might also fix the tests on iOS.
```suggestion
    [webView evaluateJavaScript:@"document.getElementsByTagName('video')[0].play()" inFrame:[webView firstChildFrame] completionHandler:nil];
    while (![webView _isPlayingAudio])
        Util::spinRunLoop();

    [webView evaluateJavaScript:@"document.querySelectorAll('iframe').forEach(iframe => iframe.remove())" completionHandler:nil];
    while ([webView _isPlayingAudio])
        Util::spinRunLoop();
```